test/unit/forge/routes/auth/index_spec.js

```diff
@@ -0,0 +1,132 @@
+
+            await expectRejection({
+                username: 'admin',
+                password: '12345678',
+                name: 'u1',
+                email: 'u1@example.com'
+            }, /invalid username/)
+
+            await expectRejection({
+                username: 'root',
+                password: '12345678',
+                name: 'u1',
+                email: 'u1@example.com'
+            }, /invalid username/)
+        })
+
+        it('rejects duplicate username', async function () {
+            app = await setup()
+            app.settings.set('user:signup', true)
+
+            await registerUser({
+                username: 'u1',
+                password: '12345678',
+                name: 'u1',
+                email: 'u1@example.com'
+            })
+
+            await expectRejection({
+                username: 'u1',
+                password: '12345678',
+                name: 'u1.2',
+                email: 'u1-2@example.com'
+            }, /username must be unique/)
+        })
+        it('rejects duplicate email', async function () {
+            app = await setup()
+            app.settings.set('user:signup', true)
+
+            await registerUser({
+                username: 'u1',
+                password: '12345678',
+                name: 'u1',
+                email: 'u1@example.com'
+            })
+
+            await expectRejection({
+                username: 'u1-2',
+                password: '12345678',
+                name: 'u1.2',
+                email: 'u1@example.com'
+            }, /email must be unique/)
+        })
+
+        it('limits how many users can be created according to license', async function () {
+            // This license has limit of 5 users (1 created by default test setup (test/unit/forge/routes/setup.js))
+            const license = 'eyJhbGciOiJFUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJGbG93Rm9yZ2UgSW5jLiIsInN1YiI6IkZsb3dGb3JnZSBJbmMuIERldmVsb3BtZW50IiwibmJmIjoxNjYyNTA4ODAwLCJleHAiOjc5ODY5ODg3OTksIm5vdGUiOiJEZXZlbG9wbWVudC1tb2RlIE9ubHkuIE5vdCBmb3IgcHJvZHVjdGlvbiIsInVzZXJzIjo1LCJ0ZWFtcyI6NTAsInByb2plY3RzIjo1MCwiZGV2aWNlcyI6NTAsImRldiI6dHJ1ZSwiaWF0IjoxNjYyNTQ4NjAyfQ.vvSw6pm-NP5e0NUL7yMOG-w0AgB8H3NRGGN7b5Dw_iW5DiIBbVQ4HVLEi3dyy9fk7WgKnloiCCkIFJvN79fK_g'
+            app = await setup({ license })
+            app.settings.set('user:signup', true)
+            // Register 4 more users to reach the limit
+            for (let i = 1; i <= 4; i++) {
+                const resp = await registerUser({
+                    username: `u${i}`,
+                    password: '12345678',
+                    name: `u${i}`,
+                    email: `u${i}@example.com`
+                })
+                resp.statusCode.should.equal(200)
+            }
+            await expectRejection({
+                username: 'u6',
+                password: '12345678',
+                name: 'u6',
+                email: 'u6@example.com'
+            }, /license limit reached/)
```

Comment:
A little puzzled here.

> This license has limit of 5 users (1 created by default test setup (test/unit/forge/routes/setup.js))

| user | count |
| - | - |
| default user | 1 |

> // Register 4 more users to reach the limit
> `for (let i = 1; i <= 4; i++) {` 

Then `for (let i = 1; i <= 4; i++) {` adds four more named ```username: `u${i}\` ```
| user | count |
| - | - |
| default user | 1 |
| u1 | 2 |
| u2 | 3 |
| u3 | 4 |
| u4 | 5 |

At this point there are 5 users.

> ```
>             await expectRejection({
>                 username: 'u6',
> ```


Where does user 'u6' get created?

Reply:
My point is I can not even see how or where there is even an attempt to create a `u6`

> Where does user 'u6' get created?

Reply:
or user `u5` for that matter (the for loop only generates u1~u4)

Reply:
The test never attempts to create `u5`... lets just call that an off-by-one error. The important part is the actual number of users created.

The function `expectRejection` wraps the call to create the user and also verifies it fails properly.

Reply:
ok, now it makes sense. 

Apart from the non descriptive function naming  of `expectRejection` and the off by one error on "u1" ~ "u4" , I now see this will work as expected.

Feel free to mark this comment as resolved however for the sake of future comprehension, I would recommend altering the loop to generate `user${i+1}`  (i.e. user2, user3...user5) and would also recommend renaming  `expectRejection` to something like `addUserExpectRejection({ username: 'user6',...`